game_service.py

```
@@ -1,54 +1,43 @@
-from queue import Queue
 import time
 import threading
 import requests
 import base64
 import json
-from constants import key_map, get_random_key
+import re
+import random
+
+from queue import Queue
+from constants import key_map
 from abc import ABC, abstractmethod
 from PIL import Image
 from io import BytesIO
-import re
+from typing import Tuple
 
 
 class GameService(ABC):
     def __init__(self, command_queue: Queue, output_queue: Queue):
         self.command_queue = command_queue
-        self.output_queue = output_queue
+        self.data_queue = output_queue
+        self._time_last_command = 0
 
     def start_game(self):
-        threading.Thread(target=self.process_output, daemon=True).start()
-        threading.Thread(target=self.read_input, daemon=True).start()
-        threading.Thread(target=self.simulate_agent, daemon=True).start()
-
-    @abstractmethod
-    def read_input(self):
-        raise NotImplementedError("Method not implemented")
-
-    @abstractmethod
-    def process_output(self):
-        raise NotImplementedError("Method not implemented")
+        threading.Thread(target=self.run_agent, daemon=True).start()
 
     @abstractmethod
     def parse_command(self, output):
         raise NotImplementedError("Method not implemented")
 
     @abstractmethod
-    def simulate_agent(self):
+    def run_agent(self):
         raise NotImplementedError("Method not implemented")
 
 
 class HTTPGameService(GameService):
-    def __init__(self, command_queue: Queue, output_queue: Queue):
+    def __init__(self, command_queue: Queue, output_queue: Queue, url: str = "http://localhost:8000/chat"):
```

Imo we should be fetching the URL as a config parameter and raise an error if it's not found, instead of having default addresses in our code

---

Agreed, common issue in the codebase as is. We need to look into builders or something similar at a later point